knebind/solver/solver.go

```diff
@@ -115,16 +164,36 @@ func Solve(tb *opb.Testbed, topo *tpb.Topology) (*binding.Reservation, error) {
 		addIntf(link.GetANode(), link.GetAInt())
 		addIntf(link.GetZNode(), link.GetZInt())
 	}
-
 	// Build the node => links map for topology
 	for _, l := range topo.GetLinks() {
 		srcNode := name2Node[l.GetANode()]
 		dstNode := name2Node[l.GetZNode()]
 		srcIntf := s.node2Intfs[srcNode][l.GetAInt()]
 		dstIntf := s.node2Intfs[dstNode][l.GetZInt()]
-		nLink := &nodeLink{srcNode: srcNode, srcIntf: srcIntf, dstNode: dstNode, dstIntf: dstIntf}
-		s.node2Links[srcNode][dstNode] = append(s.node2Links[srcNode][dstNode], nLink)
-		s.node2Links[dstNode][srcNode] = append(s.node2Links[dstNode][srcNode], nLink)
+		group := srcNode.Interfaces[l.GetAInt()].GetGroup()
+		if group == "" {
+			group = dstNode.Interfaces[l.GetZInt()].GetGroup()
+		}
+		nodeLinkPtr := &nodeLink{srcNode: srcNode, dstNode: dstNode, srcIntf: srcIntf, dstIntf: dstIntf, group: group}
+		if _, ok := s.node2Links[srcNode][dstNode]; !ok {
+			pgPtr := &nodePG{groupMap: make(map[string][]*nodeLink), sortedNames: []string{}, totalLinks: 0}
```

Comment:
no need to specify sortedNames and totalLinks. Go style is to let slices initialize to nil and integers will default to zero already.